### PR TITLE
Reenable `state` support for remote REST QPU platforms

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

Re-enable state argument support for remote REST QPU:

- Restore `replace-state-with-kernel` and `state-prep` passes. Add back the corresponding Python and C++ tests.

Note: For `state-prep` pass only support state amplitudes provided as arrays, not a `state` constructed outside the kernel with an amplitude array (i.e., state constructed by a different simulator). An error will be raised accordingly.

- Modify the Python and C++ bridges to directly handle `qvector` allocation with amplitude arrays. We no longer want to wrap it via a `state` construction (which needs to use the local simulator). 

Original implementation of these passes is in #1891 and #2291.

### CODE REVIEW

File: python/runtime/cudaq/platform/py_alt_launch_kernel.cpp [hunk 1]
```diff
@@ -405,9 +405,30 @@ void cudaq::packArgs(OpaqueArguments &argData, py::list args,
...
-            });
+            auto *stateArg = arg.cast<state *>();
+
+            if (stateArg == nullptr)
+              throw std::runtime_error("Null cudaq::state* argument passed.");
+            auto simState = cudaq::state_helper::getSimulationState(
+                const_cast<cudaq::state *>(stateArg));
+            if (!simState)
+              throw std::runtime_error("Error: Unable to retrieve simulation "
+                                       "state from cudaq::state.");
```
Comment (sacpis) [inline python/runtime/cudaq/platform/py_alt_launch_kernel.cpp]: nit: We can add `The state object may be invalid or corrupted`?

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    // the position where the lth and (l + 1)th bit strings g[l] and g[l+1] of
+    // the binary reflected Gray code differ.
+    auto position = std::log2(code[i] ^ code[(i + 1) % code.size()]);
+    // N.B: The algorithm expects the least significant bit (LSb) on the left
+    //
+    //  lsb -v
+    //       001
+    //         ^- msb
+    //
+    // Meaning that the bitstring 001 represents the number four instead of one.
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp:55]: nit: Should we add a unit test for qubit ordering to ensure that this convention is maintained correctly? We have had issue in the past with respect to qubit ordering.

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+    // We have the converted position: 2, which is what we need.
+    indices.emplace_back(numBits - position - 1);
+  }
+  return indices;
+}
+
+std::vector<double> convertAngles(const std::span<double> alphas) {
+  // Implements Eq. (3) from https://arxiv.org/pdf/quant-ph/0407010.pdf
+  //
+  // N.B: The paper does fails to explicitly define what is the dot operator in
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp]: nit: The paper fails to

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+  // the exponent of -1. Ref. 3 solves the mystery: its the bitwise inner
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp]: Sorry Ref. 3 from where? Is it from the `State preparation` section of the paper?

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+      numerator +=
+          std::pow(std::abs(data[(2 * j - 1) * (1 << (k - 1)) + l - 1]), 2);
+    }
+
+    double denominator = 0;
+    for (std::size_t l = 1; l <= (1ULL << k); ++l) {
+      denominator += std::pow(std::abs(data[(j - 1) * (1 << k) + l - 1]), 2);
+    }
+
+    if (denominator == 0.0) {
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp:139]: I think we should never enter in this loop, right? If we do, then can we throw a runtime_error, `divide by 0`?

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+      assert(numerator == 0.0 &&
+             "If the denominator is zero, the numerator must also be zero.");
+      angles.push_back(0.0);
+      continue;
+    }
+    angles.push_back(2.0 * std::asin(std::sqrt(numerator / denominator)));
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp]: If numerator / denominator > 1.0 due to floating-point errors, std::asin will return NaN.

We can do

```
double ratio = numerator / denominator;
ratio = std::min(1.0, std::max(0.0, ratio));
angles.push_back(2.0 * std::asin(std::sqrt(ratio)));
```

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+
+  std::unordered_map<std::size_t, mlir::Value> qubitRefs =
+      std::unordered_map<std::size_t, mlir::Value>();
+};
+
+class StateDecomposer {
+public:
+  StateDecomposer(StateGateBuilder &b, std::span<std::complex<double>> a,
+                  double t)
+      : builder(b), amplitudes(a), numQubits(log2(a.size())),
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp:202]: If a.size() is not a power of 2, log2 returns a non-integer. Should we validate that?

```
a,size() == 0|| (a.size() & (a.size() - 1)) != 0
```

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+    // Apply uniformly controlled z-rotations, the construction in Eq. (4).
+    for (std::size_t j = 1; j <= numQubits; ++j) {
+      auto k = numQubits - j + 1;
+      auto numControls = j - 1;
+      auto target = j - 1;
+      auto alphaZk = cudaq::details::getAlphaZ(phases, numQubits, k);
+      if (alphaZk.empty())
+        continue;
+      applyRotation<quake::RzOp>(alphaZk, numControls, target);
+    }
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp]: Seems like this code is repeated (line 222-228) apart from `alphaYk` and `alphaZk`. Should we create a function instead?

File: lib/Optimizer/Transforms/StatePreparation.cpp [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+        data = cast.getValue();
+
+      if (auto addr = data.getDefiningOp<cudaq::cc::AddressOfOp>()) {
+        auto globalName = addr.getGlobalName();
+        auto module = init->getParentOfType<mlir::ModuleOp>();
+        auto symbol = module.lookupSymbol(globalName);
+        if (auto global = dyn_cast<cudaq::cc::GlobalOp>(symbol)) {
+
+          // Read state initialization data from the global array.
+          auto vec = cudaq::opt::factory::readGlobalConstantArray(global);
```
Comment (sacpis) [inline lib/Optimizer/Transforms/StatePreparation.cpp:369]: Should we validate vec size to be a power of 2?

File: python/runtime/cudaq/algorithms/py_state.cpp [hunk 10]
```diff
@@ -172,19 +174,17 @@ class PyQPUState : public QPUState {
...
-  auto kernelName = kernel.attr("uniqName").cast<std::string>();
-  auto kernelMod = kernel.attr("qkeModule").cast<MlirModule>();
+state pyGetStateQPU(const std::string &kernelName, MlirModule kernelMod,
+                    py::args args) {
+  auto moduleOp = unwrap(kernelMod);
+  std::string mlirCode;
+  llvm::raw_string_ostream outStr(mlirCode);
+  mlir::OpPrintingFlags opf;
+  opf.enableDebugInfo(/*enable=*/true, /*pretty=*/false);
+  moduleOp.print(outStr, opf);
```
Comment (sacpis) [inline python/runtime/cudaq/algorithms/py_state.cpp:184]: Let's put the serialization code in a try/catch block. In case it fails.

File: runtime/common/ArgumentConversion.cpp [hunk 11]
```diff
@@ -348,6 +348,9 @@ static Value genConstant(OpBuilder &builder, const cudaq::state *v,
   auto loc = builder.getUnknownLoc();
   auto simState =
       cudaq::state_helper::getSimulationState(const_cast<cudaq::state *>(v));
+  if (!simState)
+    throw std::runtime_error(
+        "Error: Unable to retrieve simulation state from cudaq::state.");
```
Comment (sacpis) [inline runtime/common/ArgumentConversion.cpp]: nit: let's add `The state may be invalid or uninitialized.`?

File: runtime/common/ArgumentConversion.cpp [hunk 12]
```diff
@@ -359,8 +362,144 @@ static Value genConstant(OpBuilder &builder, const cudaq::state *v,
...
+
+  if (simState->getKernelInfo().has_value()) {
+    auto [calleeName, code, calleeArgs] = simState->getKernelInfo().value();
+    std::string calleeKernelName =
+        cudaq::runtime::cudaqGenPrefixName + calleeName;
+
+    auto ctx = builder.getContext();
+    auto loc = builder.getUnknownLoc();
+
+    assert(!code.empty() && "Quake code not found for callee");
```
Comment (sacpis) [inline runtime/common/ArgumentConversion.cpp]: Seems like `assert` is removed in release builds. Let's replace all asserts with `if .. throw std::runtime_error` logic.

File: runtime/common/ArgumentConversion.cpp [hunk 13]
```diff
@@ -359,8 +362,144 @@ static Value genConstant(OpBuilder &builder, const cudaq::state *v,
...
+    auto initKernelName = cudaq::runtime::cudaqGenPrefixName + initName;
+    auto numQubitsKernelName =
+        cudaq::runtime::cudaqGenPrefixName + numQubitsName;
+
+    // Create `callee.init_N` and `callee.num_qubits_N` used to replace
+    // `quake.materialize_state` in ReplaceStateWithKernel pass
+    if (!converter.isRegisteredKernel(initName) ||
+        !converter.isRegisteredKernel(numQubitsName)) {
+      createInitFunc(builder, substMod, calleeFunc, initKernelName);
+      createNumQubitsFunc(builder, substMod, calleeFunc, numQubitsKernelName);
```
Comment (sacpis) [inline runtime/common/ArgumentConversion.cpp]: Should we just create one when that one is missing? Wondering why do we need to create both if any one is missing?

```
if (!converter.isRegisteredKernel(initName))
    createInitFunc(builder, substMod, calleeFunc, initKernelName);
if (!converter.isRegisteredKernel(numQubitsName))
    createNumQubitsFunc(builder, substMod, calleeFunc, numQubitsKernelName);
```

File: runtime/cudaq/algorithms/get_state.h [hunk 14]
```diff
@@ -130,13 +130,17 @@ auto get_state(QuantumKernel &&kernel, Args &&...args) {
   // Store kernel name and arguments for quantum states.
   return state(
       new QPUState(details::getKernelName(std::forward<QuantumKernel>(kernel)),
+                   cudaq::get_quake_by_name(details::getKernelName(
+                       std::forward<QuantumKernel>(kernel))),
                    std::forward<Args>(args)...));
```
Comment (sacpis) [inline runtime/cudaq/algorithms/get_state.h]: Instead of calling `getKernelName` twice, we can cache it outside.

```
auto kernelName = details::getKernelName(std::forward<QuantumKernel>(kernel));
return state(
    new QPUState(kernelName,
                 cudaq::get_quake_by_name(kernelName),
                 std::forward<Args>(args)...));
```

File: runtime/cudaq/algorithms/get_state.h [hunk 15]
```diff
@@ -130,13 +130,17 @@ auto get_state(QuantumKernel &&kernel, Args &&...args) {
...
 
 #elif defined(CUDAQ_QUANTUM_DEVICE)
   // Kernel builder is MLIR-based kernel.
   if constexpr (has_name<QuantumKernel>::value)
     return state(new QPUState(
         details::getKernelName(std::forward<QuantumKernel>(kernel)),
+        cudaq::get_quake_by_name(
+            details::getKernelName(std::forward<QuantumKernel>(kernel))),
         std::forward<Args>(args)...));
```
Comment (sacpis) [inline runtime/cudaq/algorithms/get_state.h]: Same here.

File: runtime/cudaq/algorithms/get_state.h [hunk 16]
```diff
@@ -130,13 +130,17 @@ auto get_state(QuantumKernel &&kernel, Args &&...args) {
   // Store kernel name and arguments for quantum states.
   return state(
       new QPUState(details::getKernelName(std::forward<QuantumKernel>(kernel)),
+                   cudaq::get_quake_by_name(details::getKernelName(
```
Comment (sacpis) [inline runtime/cudaq/algorithms/get_state.h]: Seems like `get_quake_by_name` can return empty strings for kernels without registered MLIR (from line 116).

Let's pull out quakeCode as well and validate it?

```
auto quakeCode = cudaq::get_quake_by_name(kernelName);
if (quakeCode.empty())
    throw std::runtime_error(...);
```

File: runtime/cudaq/algorithms/get_state.h
Comment (sacpis) [inline runtime/cudaq/algorithms/get_state.h:129]: This (condition on line 129 and line 137) can be simplified as

```
#elif defined(CUDAQ_QUANTUM_DEVICE)
  #ifdef CUDAQ_LIBRARY_MODE
    if constexpr (!has_name<QuantumKernel>::value)
      throw std::runtime_error(
          "could not create state in library mode for quantum devices");
  #endif
  
  auto kernelName = details::getKernelName(std::forward<QuantumKernel>(kernel));
  auto quakeCode = cudaq::get_quake_by_name(kernelName);
  if (quakeCode.empty())
    throw std::runtime_error(
        "Cannot create state: MLIR code not found for kernel " + kernelName);
  return state(new QPUState(kernelName, quakeCode, std::forward<Args>(args)...));
#endif
```

File: runtime/cudaq/qis/qvector.h [hunk 17]
```diff
@@ -37,6 +37,54 @@ class qvector {
...
   /// @endcond
 
+  /// @brief Construct a `qvector` from an input state vector.
+  /// The number of qubits is determined by the size of the input vector.
+  /// If `validate` is set, it will check the norm of input state vector.
+  explicit qvector(const std::vector<complex> &vector, bool validate)
+      : qudits(std::log2(vector.size())) {
+    if (Levels == 2) {
+      auto numElements = std::log2(vector.size());
+      if (std::floor(numElements) != numElements)
```
Comment (sacpis) [inline runtime/cudaq/qis/qvector.h]: This comparison might fail due to floating point errors with `std::log2(vector.size())` call as it might return 2.9999999999 or 3.0000000001.
I think we can check the power of 2 using

```
vector.size() == 0 || (vector.size() & (vector.size() - 1)) != 0)
```

File: runtime/cudaq/qis/qvector.h [hunk 18]
```diff
@@ -37,6 +37,54 @@ class qvector {
   qvector() : qudits(1) {}
   /// @endcond
 
+  /// @brief Construct a `qvector` from an input state vector.
+  /// The number of qubits is determined by the size of the input vector.
+  /// If `validate` is set, it will check the norm of input state vector.
+  explicit qvector(const std::vector<complex> &vector, bool validate)
+      : qudits(std::log2(vector.size())) {
+    if (Levels == 2) {
```
Comment (sacpis) [inline runtime/cudaq/qis/qvector.h:45]: Just to confirm. We are not supporting higher level systems, right?

File: runtime/cudaq/qis/qvector.h [hunk 19]
```diff
@@ -37,6 +37,54 @@ class qvector {
...
+    getExecutionManager()->initializeState(targets, vector.data(), precision);
+  }
+  qvector(const std::vector<complex> &vector)
+      : qvector(vector, /*validate=*/false){};
+
+  qvector(const std::vector<double> &vector)
+      : qvector(std::vector<complex>{vector.begin(), vector.end()}) {}
+  qvector(std::vector<double> &&vector)
+      : qvector(std::vector<complex>{vector.begin(), vector.end()}) {}
+  qvector(const std::initializer_list<double> &list)
```
Comment (sacpis) [inline runtime/cudaq/qis/qvector.h:76]: Seems like the move constructor does not actually move. It just takes the rvalue reference but copies elements into new vector anyway. 
I think we can remove it if we don't need it or implement the move as below

```
qvector(std::vector<double> &&vector) {
    std::vector<complex> complexVec;
    complexVec.reserve(vector.size());
    for (auto &val : vector)
        complexVec.emplace_back(val, 0.0);
    *this = qvector(std::move(complexVec));
}
```

File: runtime/cudaq/qis/qvector.h [hunk 20]
```diff
@@ -37,6 +37,54 @@ class qvector {
...
+    for (auto &q : qudits)
+      targets.emplace_back(QuditInfo{Levels, q.id()});
+
+    auto precision = std::is_same_v<complex::value_type, float>
+                         ? simulation_precision::fp32
+                         : simulation_precision::fp64;
+    getExecutionManager()->initializeState(targets, vector.data(), precision);
+  }
+  qvector(const std::vector<complex> &vector)
+      : qvector(vector, /*validate=*/false){};
```
Comment (sacpis) [inline runtime/cudaq/qis/qvector.h:70]: Curious, why have we disabled validation be default?

File: targettests/execution/state_preparation_vector.cpp [hunk 21]
```diff
@@ -54,8 +64,7 @@ __qpu__ void test_real_constant_array() {
 }
 
 __qpu__ void test_real_constant_array_floating_point() {
-  cudaq::qvector v(
-      cudaq::state{std::vector<cudaq::real>({M_SQRT1_2, M_SQRT1_2, 0., 0.})});
+  cudaq::qvector v(std::vector<cudaq::real>({M_SQRT1_2, M_SQRT1_2, 0., 0.}));
```
Comment (sacpis) [inline targettests/execution/state_preparation_vector.cpp:67]: Should we add tests for states with negative, imaginary etc phases?

**Review by sacpis — APPROVED** [review]
Overall LGTM. Thanks @1tnguyen. Left a few comments.

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 2 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp:55]: This is tested in `targettests/execution/state_preparation_vector_sizes.cpp`, where we checked every amplitude location up to 5 qubits (vector size = 32).

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 5 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp:139]: I think the algorithm is expected to handle this condition (zero denominator) as a zero-rotation angle.

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 7 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp:202]: I addressed the comment [below](https://github.com/NVIDIA/cuda-quantum/pull/3824/changes#r2752072231), which added a validation check at a higher-level, before we start the decomposition.

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 9 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp:369]: Indeed, I think we should. Added in https://github.com/NVIDIA/cuda-quantum/pull/3824/changes/1c7556e548d9508996ba5ff5cb15b62839e7d17c

File: python/runtime/cudaq/algorithms/py_state.cpp (hunk 10 above)
Comment (1tnguyen) [reply python/runtime/cudaq/algorithms/py_state.cpp:184]: This IR printing procedure is not expected to fail, given that we have a valid MLIR module.

File: runtime/cudaq/algorithms/get_state.h
Comment (1tnguyen) [reply runtime/cudaq/algorithms/get_state.h:129]: I don't think that the suggested refactor is equivalent to the code. Specifically, kernel builder kernels would have MLIR code regardless of compilation mode, and thus can still be supported.

File: runtime/cudaq/qis/qvector.h (hunk 18 above)
Comment (1tnguyen) [reply runtime/cudaq/qis/qvector.h:45]: Actually, we do, e.g., 

https://github.com/NVIDIA/cuda-quantum/blob/745f0577e5a3f3848b50b478367db3cd554a200f/unittests/photonics/PhotonicsTester.cpp#L32

File: runtime/cudaq/qis/qvector.h (hunk 20 above)
Comment (1tnguyen) [reply runtime/cudaq/qis/qvector.h:70]: I'm not 100% sure, but from a high-level standpoint, this norm check procedure here is a simple one (CPU-based, no multi-threading, etc.); thus may not be suitable as a default checker. This data will eventually be sent on to the simulator, which might have a more efficient checking mechanism if it needs to validate the state vector.

File: runtime/cudaq/qis/qvector.h (hunk 19 above)
Comment (1tnguyen) [reply runtime/cudaq/qis/qvector.h:76]: This `qvector` class is a representation of the qubit register, _not_ a state vector, hence we don't store a `complexVec` in this class. For simulation, the data is passed (as a data pointer + size) to the simulator to create its internal data structure.

File: targettests/execution/state_preparation_vector.cpp (hunk 21 above)
Comment (1tnguyen) [reply targettests/execution/state_preparation_vector.cpp:67]: I added a test for a negative amplitude value in https://github.com/NVIDIA/cuda-quantum/pull/3824/changes/1c7556e548d9508996ba5ff5cb15b62839e7d17c. 

A complex value (with imaginary part) amplitude test would be quite tricky for hardware QPUs, where we can only sample the bitstrings.

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 4 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp]: I need to read the paper to figure out that it's referring to the ref [3] in the above paper. I updated the comment with the explicit arxiv link for that ref [3], along with the equation number.

File: python/runtime/cudaq/platform/py_alt_launch_kernel.cpp (hunk 1 above)
Comment (1tnguyen) [reply python/runtime/cudaq/platform/py_alt_launch_kernel.cpp]: Fixed the error message.

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 6 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp]: Fixed as suggested.

File: lib/Optimizer/Transforms/StatePreparation.cpp (hunk 8 above)
Comment (1tnguyen) [reply lib/Optimizer/Transforms/StatePreparation.cpp]: Good point. Refactored this as a helper.

File: runtime/common/ArgumentConversion.cpp (hunk 11 above)
Comment (1tnguyen) [reply runtime/common/ArgumentConversion.cpp]: Fixed the error message.

File: runtime/common/ArgumentConversion.cpp (hunk 13 above)
Comment (1tnguyen) [reply runtime/common/ArgumentConversion.cpp]: Definitely. Split the two conditions accordingly.

File: runtime/common/ArgumentConversion.cpp (hunk 12 above)
Comment (1tnguyen) [reply runtime/common/ArgumentConversion.cpp]: Fixed as suggested.

File: runtime/cudaq/algorithms/get_state.h (hunk 14 above)
Comment (1tnguyen) [reply runtime/cudaq/algorithms/get_state.h]: Sure, fixed accordingly.

File: runtime/cudaq/algorithms/get_state.h (hunk 15 above)
Comment (1tnguyen) [reply runtime/cudaq/algorithms/get_state.h]: Fixed as well.

File: runtime/cudaq/algorithms/get_state.h (hunk 16 above)
Comment (1tnguyen) [reply runtime/cudaq/algorithms/get_state.h]: Here, we don't pass `false` to `get_quake_by_name`, so it would throw if it cannot find the quake code.

File: runtime/cudaq/qis/qvector.h (hunk 17 above)
Comment (1tnguyen) [reply runtime/cudaq/qis/qvector.h]: Fixed as suggested.
